src/components/DataTable/DataTable.js

```diff
@@ -1,403 +1,374 @@
+  static propTypes = {
+    /**
+     * The `rows` prop is where you provide us with a list of all the rows that
+     * you want to render in the table. The only hard requirement is that this
+     * is an array of objects, and that each object has a unique `id` field
+     * available on it.
+     */
+    rows: PropTypes.arrayOf(
+      PropTypes.shape({
+        id: PropTypes.string.isRequired,
+      })
+    ),
+
+    /**
+     * The `headers` prop represents the order in which the headers should
+     * appear in the table. We expect an array of objects to be passed in, where
+     * `key` is the name of the key in a row object, and `header` is the name of
+     * the header.
+     */
+    headers: PropTypes.arrayOf(
+      PropTypes.shape({
+        key: PropTypes.string.isRequired,
+        header: PropTypes.string.isRequired,
+      })
+    ).isRequired,
+
+    /**
+     * Provide an `onChange` hook where you can subscribe to internal state
+     * changes of the component.
+     */
+    onChange: PropTypes.func,
+
+    /**
+     * Optional hook to manually control sorting of the rows.
+     */
+    sortRow: PropTypes.func,
+
+    /**
+     * Optional hook to manually control filtering of the rows from the
+     * TableToolbarSearch component
+     */
+    filterRows: PropTypes.func,
+
+    /**
+     * Provide a string for the current locale
+     */
+    locale: PropTypes.string,
+  };
 
-const sortRow = (key, direction) => (a, b) => {
-  if (direction === 'DESC') {
-    return a[key].localeCompare(b[key]);
-  }
+  static defaultProps = {
+    sortRow: defaultSortRow,
+    filterRows: defaultFilterRows,
+    locale: 'en',
+  };
 
-  return b[key].localeCompare(a[key]);
-};
+  /**
+   * Static helper to derive the next state from the given props and the
+   * prevState. Potential future-facing API hook for React v17.
+   *
+   * Currently, it's being used as a way to normalize the incoming data that we
+   * are receiving for rows
+   */
+  static getDerivedStateFromProps = props => {
+    const { rowIds, rowsById, cellsById } = normalize(
+      props.rows,
+      props.headers
+    );
+    return {
+      rowIds,
+      rowsById,
+      cellsById,
+      sortDirection: initialSortState,
+      sortHeaderKey: null,
+      // Copy over rowIds so the reference doesn't mutate the stored
+      // `initialRowOrder`
+      initialRowOrder: rowIds.slice(),
+
+      filterInputValue: null,
+
+      // Optionally state field to indicate whether a consumer should show a
+      // batch actions menu
+      shouldShowBatchActions: false,
+    };
+  };
 
-export class DataTable extends Component {
   constructor(props) {
     super(props);
-    this.state = {
-      rows: props.initialRows,
-      sortKey: null,
-      sortDirection: null,
-    };
-    this.sortBy = this.sortBy.bind(this);
-    this.getHeaderProps = this.getHeaderProps.bind(this);
+    this.state = DataTable.getDerivedStateFromProps(props);
```

Comment:
While `Class.method()` works, `this.constructor.method()` is the way of calling static method - More robust way against change in class name, etc.